storage/sql/private.go

```diff
@@ -22,10 +22,19 @@ func (s *Storage) InsertOrUpdatePrivateXML(privateXML []xmpp.XElement, namespace
 	}
 	rawXML := buf.String()
 
+	var suffix string
+
+	switch s.engine {
+	case "mysql":
+		suffix = "ON DUPLICATE KEY UPDATE data = ?, updated_at = NOW()"
+	case "postgresql":
+		suffix = "ON CONFLICT (username) DO UPDATE SET data = $4, updated_at = NOW()"
```

Comment:
Agreed, this is a much better approach, but first we'd need to separate mysql/psql storage handler, since now they share vast majority of the code, so engine-specific things are messy to implement. This should be the goal, though.